hnix-store-core/src/System/Nix/Internal/Path.hs

```diff
@@ -0,0 +1,150 @@
+  name = pathName . T.drop 1 . snd $ parts
+
+
+-- experimental
+-- Directory of the store
+type StoreDir = Text
+type Stored a = (StoreDir, a)
+
+-- wrap StoreDir and Path into tuple
+makeStored :: StoreDir -> Path -> Stored Path
+makeStored sl p = (sl, p)
+
+type PathType = Text
+-- "text:<r1>:<r2>:...<rN>"
+-- "source"
+-- "output:<id>"
+--    <id> is the name of the output (usually, "out").
+
+-- store settings
+data Settings = Settings {
+  storeDir :: StoreDir -- settings.nixStore
+  } deriving (Eq, Show)
+
+-- build a store path in the following form:
+-- <storeDir>/<hash>-<pathName>
+storedToText :: Stored Path -> Text
+storedToText (storeLoc, (Path digest pName)) = T.concat
+  [ storeLoc
+  , "/"
+  , printAsBase32 @PathHashAlgo digest
+  , "-"
+  , pathNameContents pName
+  ]
+
+makeStorePath :: (HasDigest a) => PathType -> PathName -> Digest a -> Settings -> Text
+makeStorePath typ pName digest settings = T.concat
+  [ storeDir settings
+  , "/"
+  , printAsBase32 @PathHashAlgo $ pathHash typ pName digest (storeDir settings)
+  , "-"
+  , pathNameContents pName
+  ]
+
+makeStorePath' :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Path
+makeStorePath' typ pName digest storeLoc = snd $ makeStoredPath typ pName digest storeLoc
+
+-- | build Stored Path from the type of the path, path name and a digest stored at StoreDir
+-- As StoreDir is part of the path hashing process we need to take it into account
+-- when building Path(s)
+makeStoredPath :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Stored Path
+makeStoredPath typ pName digest storeLoc = makeStored storeLoc $ Path (pathHash typ pName digest storeLoc) pName
+
+-- build <h> string which is a truncated base32 formatted SHA256 hash of <s>
+pathHash :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Digest PathHashAlgo
+pathHash typ pName digest storeLoc = System.Nix.Hash.hash . BSC.pack . T.unpack $
+  makePathDigestString typ pName digest storeLoc
+
+-- build <s> string which is hashed and used in makeStorePath
+-- <s> = "<pathType>:<hash_algo>:<base16_hash>:<storeDir>:<pathName>"
+-- (exposed for testing purposes only)
+makePathDigestString ::  (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Text
+makePathDigestString typ pName digest storeLoc = T.intercalate (T.pack ":")
+  [ typ
+  , printHashAlgo digest
+  , printAsBase16 digest
+  , storeLoc
+  , pathNameContents pName
+  ]
+
+-- make output path from `PathName` digest and outputId which typically is "out"
+makeOutputPath :: (HasDigest a) => PathName -> Digest a -> T.Text -> Settings -> Text
+makeOutputPath pName digest outputId settings =
+  makeStorePath typ (adjustName pName) digest settings
+  where
+    typ = T.concat [ "output:", outputId ]
+    adjustName n | outputId == "out" = n
+    adjustName (PathName name) | otherwise = PathName $ T.concat [ name, T.pack "-", outputId ]
+
+type Recursive = Bool
```

Comment:
Nit: `data Recursive = Recursive | NotRecursive`  preferred over type alias.

Reply:
Actually:
```
-- | Whether the hash is computedover the serialisation of the output path
-- or over the contents of the single non-executable regular file
-- See thesis page 107
data OutputHashMode = Recursive | Flat

Reply:
This is still being broken up into manageable chunks,but see https://github.com/haskell-nix/hnix-store/commit/974a944bf5a30e1090acb8587817afa2433bd39a#diff-d6eea410e6c564641deccfa88ae99d1fR101